nucliadb/src/nucliadb/search/search/query_parser/parsers/find.py

```diff
@@ -66,20 +132,117 @@ async def parse(self) -> UnitRetrieval:
             rank_fusion.window = max(rank_fusion.window, reranker.window)
 
         return UnitRetrieval(
-            top_k=top_k,
+            query=self._query,
+            top_k=self._top_k,
+            filters=filters,
             rank_fusion=rank_fusion,
             reranker=reranker,
         )
 
-    def _parse_top_k(self) -> int:
-        assert self.item.top_k is not None, "top_k must have an int value"
-        top_k = self.item.top_k
-        return top_k
+    async def parse_relation_query(self) -> RelationQuery:
+        detected_entities = await self._get_detected_entities()
+
+        deleted_entity_groups = await self.fetcher.get_deleted_entity_groups()
+
+        meta_cache = await self.fetcher.get_entities_meta_cache()
+        deleted_entities = meta_cache.deleted_entities
+
+        return RelationQuery(
+            detected_entities=detected_entities,
+            deleted_entity_groups=deleted_entity_groups,
+            deleted_entities=deleted_entities,
+        )
+
+    async def _get_detected_entities(self) -> list[utils_pb2.RelationNode]:
+        """Get entities from request, either automatically detected or
+        explicitly set by the user."""
+
+        if self.item.query_entities:
+            detected_entities = []
+            for entity in self.item.query_entities:
+                relation_node = utils_pb2.RelationNode()
+                relation_node.value = entity.name
+                if entity.type is not None:
+                    relation_node.ntype = RelationNodeTypeMap[entity.type]
+                if entity.subtype is not None:
+                    relation_node.subtype = entity.subtype
+                detected_entities.append(relation_node)
+        else:
+            detected_entities = await self.fetcher.get_detected_entities()
+
+        meta_cache = await self.fetcher.get_entities_meta_cache()
+        detected_entities = expand_entities(meta_cache, detected_entities)
+
+        return detected_entities
+
+    async def _parse_filters(self) -> Filters:
```

Comment:
Maybe we can extract this method to a common file? I think it's the same for search & find. It could just take a `Union[_SearchParser,_QueryParser]` as parameter (or a union of `SearchRequest,FindRequest` and pass the `item`).

Only if it's straight-forward. Otherwise this is fine.